src/Ipac_Table_Writer/internals.cxx

```diff
@@ -108,94 +192,257 @@ void tablator::Ipac_Table_Writer::write_selected_records(
     }
     os << "\n|";
 
-
     // Write column types
-    for (size_t i = 1; i < columns.size(); ++i) {
-        auto column = columns[i];
-        Data_Type active_datatype = datatypes_for_writing[i];
-        size_t effective_array_size =
-                get_effective_array_size(active_datatype, column.array_size);
-
-        for (size_t element = 0; element < effective_array_size; ++element) {
-            os << std::setw(ipac_column_widths[i])
-               << tablator::Ipac_Table_Writer::to_ipac_string(active_datatype) << "|";
-        }
+    for (size_t col_id = 1; col_id < columns.size(); ++col_id) {
+        write_column_type(table, os, datatypes_for_writing, col_id,
+                          ipac_column_widths[col_id]);
     }
 
     os << "\n|";
 
     // Write column units
-    for (size_t i = 1; i < columns.size(); ++i) {
-        auto column = columns[i];
-        size_t effective_array_size =
-                get_effective_array_size(datatypes_for_writing[i], column.array_size);
-        auto unit = column.field_properties.attributes.find("unit");
-        std::string unit_str =
-                (unit == column.field_properties.attributes.end()) ? " " : unit->second;
+    for (size_t col_id = 1; col_id < columns.size(); ++col_id) {
+        size_t effective_array_size = get_effective_array_size(
+                datatypes_for_writing[col_id], columns[col_id].array_size);
+        write_column_unit(table, os, col_id, ipac_column_widths[col_id],
+                          effective_array_size);
+    }
+    os << "\n|";
 
-        for (size_t element = 0; element < effective_array_size; ++element) {
-            os << std::setw(ipac_column_widths[i]) << unit_str << "|";
-        }
+    // Write column null property
+    for (size_t col_id = 1; col_id < columns.size(); ++col_id) {
+        size_t effective_array_size = get_effective_array_size(
+                datatypes_for_writing[col_id], columns[col_id].array_size);
+        write_column_null(table, os, col_id, ipac_column_widths[col_id],
+                          effective_array_size);
+    }
+    os << "\n";
+}
+
+
+/**********************************************************/
+// The IPAC table format does not support array-valued non-char columns.
+// A array column of fixed size n named "col" of non-char type is formatted
+// as single-valued columns named "col_0", "col_1", ... "col_{n-1}".
+void tablator::Ipac_Table_Writer::write_column_headers(
+        const Table& table, std::ostream& os,
+        const std::vector<size_t>& included_column_ids,
+        const std::vector<Data_Type>& datatypes_for_writing) {
+    const std::vector<size_t> ipac_column_widths = table.get_column_widths();
+    const auto& columns = table.columns;
+    size_t total_record_width = 0;
+
+    std::vector<size_t>::const_iterator cols_begin = included_column_ids.begin();
+    if (cols_begin != included_column_ids.end() && *cols_begin == 0) {
+        ++cols_begin;
+    }
```

Comment:
This feels like it should go inside each for loop, since we aren't assuming sorted and unique column IDs - or did I miss something?